x-pack/plugins/maps/public/classes/styles/vector/properties/dynamic_color_property.tsx

```diff
@@ -145,81 +145,59 @@ export class DynamicColorProperty extends DynamicStyleProperty<ColorDynamicOptio
       if (!this._options.customColorRamp || !this._options.customColorRamp.length) {
         // custom color ramp config is not complete
         return null;
       }
-
-      const colorStops: Array<number | string> = this._options.customColorRamp.reduce(
-        (accumulatedStops: Array<number | string>, nextStop: OrdinalColorStop) => {
-          return [...accumulatedStops, nextStop.stop, nextStop.color];
-        },
-        []
-      );
-      const firstStopValue = colorStops[0] as number;
-      const lessThanFirstStopValue = firstStopValue - 1;
-      return [
-        'step',
-        ['coalesce', [this.getMbLookupFunction(), targetName], lessThanFirstStopValue],
-        RGBA_0000, // MB will assign the base value to any features that is below the first stop value
-        ...colorStops,
-      ];
-    }
-
-    if (this.getDataMappingFunction() === DATA_MAPPING_FUNCTION.PERCENTILES) {
+      colorStops = this._getCustomRampColorStops();
+    } else if (this.getDataMappingFunction() === DATA_MAPPING_FUNCTION.PERCENTILES) {
       const percentilesFieldMeta = this.getPercentilesFieldMeta();
       if (!percentilesFieldMeta || !percentilesFieldMeta.length) {
         return null;
       }
-
-      const colorStops = getPercentilesMbColorRampStops(
+      colorStops = getPercentilesMbColorRampStops(
         this._options.color ? this._options.color : null,
         percentilesFieldMeta
       );
-      if (!colorStops) {
+    } else {
+      const rangeFieldMeta = this.getRangeFieldMeta();
+      if (!rangeFieldMeta) {
         return null;
       }
-
-      const lessThanFirstStopValue = percentilesFieldMeta[0].value - 1;
-      return [
-        'step',
-        ['coalesce', [this.getMbLookupFunction(), targetName], lessThanFirstStopValue],
-        RGBA_0000,
-        ...colorStops,
-      ];
-    }
-
-    const rangeFieldMeta = this.getRangeFieldMeta();
-    if (!rangeFieldMeta) {
-      return null;
+      colorStops = getOrdinalMbColorRampStops(
+        this._options.color ? this._options.color : null,
+        rangeFieldMeta.min,
+        rangeFieldMeta.max
+      );
     }
 
-    const colorStops = getOrdinalMbColorRampStops(
-      this._options.color ? this._options.color : null,
-      rangeFieldMeta.min,
-      rangeFieldMeta.max
-    );
-    if (!colorStops) {
-      return null;
-    }
+    return colorStops && colorStops.length >= 2
+      ? [
+          'step',
+          makeMbClampedNumberExpression({
+            minValue: colorStops[0],
+            maxValue: colorStops[colorStops.length - 2],
+            lookupFunction: this.getMbLookupFunction(),
+            fallback: colorStops[0] - 1,
+            fieldName: this.getFieldName(),
+          }),
+          RGBA_0000, // MB will assign the base value to any features that is below the first stop value
+          ...colorStops,
+        ]
+      : null;
+  }
 
-    const lessThanFirstStopValue = rangeFieldMeta.min - 1;
-    return [
-      'interpolate',
```

Comment:
mapbox "interpolate" function not really needed since "getOrdinalMbColorRampStops" was already breaking the domain into ranges. With this PR, now "steps" is used